src/snapshot.rs

```diff
@@ -152,24 +191,71 @@ fn snapshot_memories<'a>(instance: &wasmtime::Instance) -> (Vec<u32>, Vec<DataSe
 
         // Only merge segments if they are contiguous or if it is definitely
         // more size efficient than leaving them apart.
-        let distance = b.offset - (a.offset + u32::try_from(a.data.len()).unwrap());
-        if distance > MIN_ACTIVE_SEGMENT_OVERHEAD {
+        let gap = a.gap(b);
+        if gap > MIN_ACTIVE_SEGMENT_OVERHEAD {
             merged_data_segments.push(*b);
             continue;
         }
 
-        // Okay, merge them together into `a` (so that the next iteration
-        // can merge it with its predecessor) and then remove `b`!
-        a.data = unsafe {
-            let distance = usize::try_from(distance).unwrap();
-            debug_assert_eq!(
-                a.data
-                    .as_ptr()
-                    .offset(isize::try_from(a.data.len() + distance).unwrap()),
-                b.data.as_ptr()
-            );
-            std::slice::from_raw_parts(a.data.as_ptr(), a.data.len() + distance + b.data.len())
-        };
+        // Okay, merge them together into `a` (so that the next iteration can
+        // merge it with its predecessor) and then omit `b`!
+        let merged_data = unsafe { a.merge(b) };
+        a.data = merged_data;
+    }
+
+    // Engines apply a limit on how many segments a module may contain, and
+    // Wizer can run afoul of it. In this case, we need to merge data segments
+    // together until our number of data segments fits within the limit.
+    if merged_data_segments.len() >= MAX_DATA_SEGMENTS {
+        // We need to remove `excess` data segments. Find the `excess` smallest
+        // gaps between the start of one segment and the next. We will merge
+        // these segments together. Because they are the smallest gaps, this
+        // will bloat the size of our data segment the least.
+        let excess = merged_data_segments.len() - MAX_DATA_SEGMENTS;
+
+        #[derive(Clone, Copy, PartialEq, Eq)]
+        struct GapIndex {
+            gap: u32,
+            index: usize,
+        }
+        impl Ord for GapIndex {
+            fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+                // NB: Bigger gaps first.
+                other
+                    .gap
+                    .cmp(&self.gap)
+                    .then_with(|| self.index.cmp(&other.index))
+            }
+        }
+        impl PartialOrd for GapIndex {
+            fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+                Some(self.cmp(other))
+            }
+        }
+
+        let mut smallest_gaps = BinaryHeap::<GapIndex>::with_capacity(excess + 1);
+        for i in 0..merged_data_segments.len() - 1 {
```

Comment:
This might be a good use case for `merged_data_segments.windows(2).enumerate()` perhaps/

Reply:
Good call.